swap/swap.go

```diff
@@ -314,22 +314,52 @@ func createOwner(prvkey *ecdsa.PrivateKey) *Owner {
 	}
 }
 
-// Add is the (sole) accounting function
-// Swap implements the protocols.Balance interface
-func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
+// checkBalance checks that the amount would not result in crossing the disconnection threshold
+func (s *Swap) checkBalance(amount int64, peer *protocols.Peer) (err error) {
 	swapPeer := s.getPeer(peer.ID())
 	if swapPeer == nil {
 		return fmt.Errorf("peer %s not a swap enabled peer", peer.ID().String())
 	}
-	swapPeer.lock.Lock()
-	defer swapPeer.lock.Unlock()
 
 	// check if balance with peer is over the disconnect threshold and if the message would increase the existing debt
 	balance := swapPeer.getBalance()
 	if balance >= s.params.DisconnectThreshold && amount > 0 {
 		return fmt.Errorf("balance for peer %s is over the disconnect threshold %d and cannot incur more debt, disconnecting", peer.ID().String(), s.params.DisconnectThreshold)
 	}
 
+	return nil
+}
+
+// Check is called as a *dry run* before applying the actual accounting to an operation.
+// It only checks that performing a given accounting operation would not incur in an error.
+// If it returns no error, this signals to the caller that the operation is safe
+func (s *Swap) Check(amount int64, peer *protocols.Peer) (err error) {
+	swapPeer := s.getPeer(peer.ID())
```

Comment:
You do `s.getPeer(peer.ID())` here and you also do it in the function `checkBalance` (`swap:319`)

Reply:
That's because `getPeer` is under a lock.

`getBalance` is called from different places (also from `Add`) and thus needs to be called independently

Reply:
I don't understand then why you would call it at this place at all. Seemingly all you are doing is verifying wether the peer is a swapPeer, but this is also the first thing you are doing in `checkBalance`. 
I think that this function can be improved by not doing a call to `getPeer` at all, or by modifying `checkBalance` to only accept a `swapPeer` and then pass the `swapPeer` to `checkBalance`

Reply:
I tried to explain before but I'll try again, a bit more explicit.

There are two things here:
* `Check` is called from the `p2p/protocols` package, which does **not** know the notion of a Swap peer, thus, `Check` needs to have a `protocols.Peer` in its signature - hence, we need a conversion
* `checkBalance` is also called from `Add`, which also for the same reason has a `protocols.Peer` in its signature, thus also needs a conversion

`checkBalance` is called independently, from two places.

The other thing is that we are locking the `swapPeer` during these operations. The lock is on the `swapPeer`, that is why we need it, 

Your optimization suggestions would be more stringent if we can remove the lock from `swapPeer` in this case as we are already locking the `protocols.Peer` in the `protocols` package, as @mortelli also noted. However, I am not 100% sure we can omit that lock, it is the same network remote peer but two different objects (`protocols` vs `swap`).  I don't feel safe doing this at the moment, which is why I suggest keeping it as-is, but it is a valid suggestion to revisit this part and analyze if the lock can be omitted, in which case we could probably optimize those methods.

Reply:
However your second suggestion is good, I changed `checkBalance` to only accept a swap `Peer`